clients/publisherlite.go

```diff
@@ -0,0 +1,102 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clients
+
+import (
+	"context"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
+	"cloud.google.com/go/pubsublite/pscompat"
+	"github.com/conduitio-labs/conduit-connector-gcp-pubsub/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"google.golang.org/api/option"
+)
+
+const topicPathFmt = "projects/%s/locations/%s/topics/%s"
+
+// A PublisherLite represents a publisher struct with a GCP Pub/Sub Lite client.
+type PublisherLite struct {
+	publisher *pscompat.PublisherClient
+
+	resultsCh chan *result
+	ctxCancel context.CancelFunc
+}
+
+// NewPublisherLite initializes a new publisher client of GCP Pub/Sub Lite.
+func NewPublisherLite(ctx context.Context, cfg config.Destination) (*PublisherLite, error) {
+	credential, err := cfg.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	topicPath := fmt.Sprintf(topicPathFmt, cfg.ProjectID, cfg.Location, cfg.TopicID)
+
+	publisher, err := pscompat.NewPublisherClientWithSettings(ctx, topicPath, pscompat.PublishSettings{
+		DelayThreshold: cfg.BatchDelay,
+		CountThreshold: cfg.BatchSize,
+	}, option.WithCredentialsJSON(credential))
+	if err != nil {
+		return nil, fmt.Errorf("create publisher client: %w", err)
+	}
+
+	cctx, cancel := context.WithCancel(ctx)
+
+	publisherLite := &PublisherLite{
+		publisher: publisher,
+		resultsCh: make(chan *result, cfg.BatchSize),
+		ctxCancel: cancel,
+	}
+
+	go func() {
+		for {
+			select {
+			case <-cctx.Done():
+				return
+			case res := <-publisherLite.resultsCh:
+				_, err := res.publishResult.Get(ctx)
+
+				err = res.ackFunc(err)
+				if err != nil {
+					sdk.Logger(ctx).Err(err).Msg("failed to call ackFunc")
```

Comment:
what about save err in  error channel? Then we can check if it chan not empty and return error

Reply:
I've added a channel to return this error and added a select to the `WriteAsync` and `Teardown` methods.